File: 151-reverse-words-in-a-string/reverse-words-in-a-string.java
class Solution {
    public void rev(char[] s, int start, int end) {
        if (start >= end)
            return;
        while (start < end) {
            char ch = s[start];
            s[start] = s[end];
            s[end] = ch;
            start++;
            end--;
        }
    }

    public String reverseWords(String s) {
        int n = s.length();

        char[] arr = s.toCharArray();

        rev(arr, 0, n - 1);

        int start = 0;
        String ans = "";

        for (int i = 0; i < n; i++) {
            if (arr[i] == ' ') {
                String tmp = new String(arr, start, i - start);
                if (tmp.length() > 0) {
                    char[] tmpArr = tmp.toCharArray();

                    rev(tmpArr, 0, tmpArr.length - 1);

                    ans = ans + " " + new String(tmpArr);
                }
                start = i + 1;
            }
        }

        String tmp = new String(arr, start, n - start);
        if (tmp.length() > 0) {
            char[] tmpArr = tmp.toCharArray();
            rev(tmpArr, 0, tmpArr.length - 1);

            ans = ans + " " + new String(tmpArr);
        }

        return ans.substring(1);
    }
}

File: 151-reverse-words-in-a-string/reverse-words-in-a-string.cpp
class Solution {
public:
    void reverse(string& s) {
        int start = 0;
        int end = s.size() - 1;
        if (start >= end) {
            return;
        }
        while (start <= end) {
            char temp = s[start];
            s[start] = s[end];
            s[end] = temp;
            start++;
            end--;
        }
    }
    string reverseWords(string s) {
        int n = s.size();
        reverse(s);

        string ans = "";
        string word = "";

        for (int i = 0; i < n; i++) {
            if (s[i] == ' ') {
                if (word.size() > 0) {
                    reverse(word);
                    ans += ' ' + word;
                    word = "";
                }
            } else {
                word += s[i];
            }
        }
        if (word.size() > 0) {
            reverse(word);
            ans += ' ' + word;
        }
        return ans.substr(1);
    }
};